---
title: "MetaPR2 - Figures"
date: "`r Sys.Date()`"
output:
  rmdformats::readthedown:
    highlight: kate
---


# Initialize

```{r init, echo=TRUE, message=FALSE, warning=FALSE}
source(here::here("R", "init.R"))
```

# Figures

## Parameters

```{r}
cluster_asv = TRUE
protist_only = TRUE
```

## Tremap for all


```{r}

fig_treemap <- asv_set$df %>% 
  treemap(asv_set$fasta, "kingdom", global$supergroup_colors)

fig_treemap_all <- fig_treemap$g 

fig_treemap_all

```


## V9 only

#### Cluster ASV =   FALSE 
Total numbers of V9 samples:  1166 
Total numbers of V9 ASVs:  32323 
Total numbers of V9 reads:  804 940 059 
#### Cluster ASV =   TRUE 
Total numbers of V9 samples:  1166 
Total numbers of V9 ASVs:  32195 
Total numbers of V9 reads:  804 940 059 


```{r}
asv_set_V9 <- list()

for (cluster_asv in c(FALSE, TRUE)) {


if (cluster_asv) {
  asv_set_V9$df <- asv_set$df_cluster 
} else {
  asv_set_V9$df <- asv_set$df 
}

asv_set_V9$df <- asv_set_V9$df %>% 
  dplyr::filter(gene_region == "V9", 
                DNA_RNA == "DNA")

n_reads_pct_V9 <- sum(asv_set_V9$df$n_reads_pct)
n_reads_V9 <- sum(asv_set_V9$df$n_reads)

samples_V9 <- asv_set_V9$df %>% 
  pull("file_code") %>% 
  unique()

n_samples_V9 <- length(samples_V9)
  
asv_code_V9 <- asv_set_V9$df %>% 
  dplyr::filter(gene_region == "V9", 
                DNA_RNA == "DNA") %>% 
  pull("asv_code") %>% 
  unique()

asv_set_V9$fasta <- asv_set$fasta %>% 
  filter(asv_code %in% asv_code_V9)

n_asvs_V9 <- nrow(asv_set_V9$fasta)

cat("#### Cluster ASV =  ",cluster_asv, "\n")
cat("Total numbers of V9 samples: ",n_samples_V9, "\n")
cat("Total numbers of V9 ASVs: ",n_asvs_V9 , "\n")
cat("Total numbers of V9 reads: ",formatC(n_reads_V9,big.mark = " ") , "\n")

}

```

## V4 - protists - DNA

#### Protist only =   FALSE 
##### Cluster ASV =   FALSE 
Total numbers of V4 samples:  2897 
Total numbers of V4 sampling sites:  846 
Total numbers of V4 ASVs:  57308 
Total numbers of V4 reads:  224 609 623 
#### Protist only =   FALSE 
##### Cluster ASV =   TRUE 
Total numbers of V4 samples:  2897 
Total numbers of V4 sampling sites:  846 
Total numbers of V4 ASVs:  48542 
Total numbers of V4 reads:  224 609 623 

#### Protist only =   TRUE 
##### Cluster ASV =   FALSE 
Total numbers of V4 samples:  2889 
Total numbers of V4 sampling sites:  842 
Total numbers of V4 ASVs:  41255 
Total numbers of V4 reads:  132 294 509 
#### Protist only =   TRUE 
##### Cluster ASV =   TRUE 
Total numbers of V4 samples:  2889 
Total numbers of V4 sampling sites:  842 
Total numbers of V4 ASVs:  33311 
Total numbers of V4 reads:  132 294 509


```{r}


for (protist_only in c(FALSE, TRUE)) {
for (cluster_asv in c(FALSE, TRUE)) {

division_memoved <- c()

if (protist_only) {
division_memoved <- c("Metazoa", "Fungi", "Streptophyta")
}

asv_set_V4 <- list()

if (cluster_asv) {
  asv_set_V4$df <- asv_set$df_cluster 
} else {
  asv_set_V4$df <- asv_set$df 
}

asv_set_V4$df <- asv_set_V4$df %>% 
  dplyr::filter(gene_region == "V4", 
                DNA_RNA == "DNA",
                !(division %in% division_memoved))

n_reads_pct_V4 <- sum(asv_set_V4$df$n_reads_pct)
n_reads_V4 <- sum(asv_set_V4$df$n_reads)

samples_V4 <- asv_set_V4$df %>% 
  pull("file_code") %>% 
  unique()

n_samples_V4 <- length(samples_V4)

sampling_sites_V4 <- asv_set_V4$df %>%
  select(latitude, longitude) %>% 
  distinct()

n_sampling_sites_V4 <- nrow(sampling_sites_V4)
  
asv_code_V4 <- asv_set_V4$df %>% 
  dplyr::filter(gene_region == "V4", 
                DNA_RNA == "DNA",
                !(division %in% division_memoved)) %>% 
  pull("asv_code") %>% 
  unique()

asv_set_V4$fasta <- asv_set$fasta %>% 
  filter(asv_code %in% asv_code_V4)

asv_set_V4$samples <- asv_set$samples %>% 
  filter(file_code %in% asv_set_V4$df$file_code)

n_asvs_V4 <- nrow(asv_set_V4$fasta)

cat("#### Protist only =  ",protist_only, "\n")
cat("##### Cluster ASV =  ",cluster_asv, "\n")
cat("Total numbers of V4 samples: ",n_samples_V4, "\n" )
cat("Total numbers of V4 sampling sites: ",n_sampling_sites_V4, "\n" )
cat("Total numbers of V4 ASVs: ",n_asvs_V4 , "\n")
cat("Total numbers of V4 reads: ", formatC(n_reads_V4,big.mark = " ") , "\n" )
  }
}
```

## Total number of species

```{r}

n <- asv_set$df %>% 
  select(species) %>% 
  distinct() %>% 
  nrow()

cat("Total # of species: ",n, "\n")

n <- asv_set_V9$df %>% 
  select(species) %>% 
  distinct() %>% 
  nrow()

cat("V9 # of species: ",n, "\n")

n <- asv_set_V4$df %>% 
  select(species) %>% 
  distinct() %>% 
  nrow()

cat("V4 # of protist species: ",n, "\n")

```

## Treemaps (for V4)

```{r}

fig_treemap <- asv_set_V4$df %>% 
  treemap(asv_set_V4$fasta, "kingdom", global$supergroup_colors)

fig_treemap <- fig_treemap$g 

fig_treemap

```

## Treemaps (for V4) for 454

```{r}

datasets_454 <- asv_set$datasets %>% 
  filter(str_detect(sequencing_technology, "454")) %>% 
  pull(dataset_id)

fig_treemap_454 <- asv_set_V4$df %>% 
  filter(dataset_id %in% datasets_454) %>% 
  treemap(asv_set_V4$fasta, "kingdom", global$supergroup_colors)

fig_treemap_454 <- fig_treemap_454$g 

fig_treemap_454

```

## Analysis at genus level



### Most abundant genera

```{r}

genus_removed <- "_X|clade|NASSO"

reads <- asv_set_V4$df %>% 
  mutate(n_reads = n_reads_pct/n_reads_pct_V4*100) %>%
  filter(!str_detect(genus, genus_removed)) %>% 
  phyloseq_long_bargraph (n_bars=30, title="", text_scaling = 0.75,
                                   use_asv = FALSE,
                                   taxo_level= genus,
                                   taxo_level_fill = supergroup,
                                   taxo_colors_fill = global$supergroup_color)
 gg_reads <- reads$gg  + 
  ylab("% of total reads") +
  ggtitle("Most abundant genera") 

stations <- asv_set_V4$df %>% 
  select(file_code, supergroup, genus) %>% 
  distinct() %>% 
  group_by(supergroup, genus) %>% 
  count() %>% 
  mutate(n_reads = n/n_samples_V4*100) %>%
  filter(!str_detect(genus, genus_removed)) %>% 
  phyloseq_long_bargraph (n_bars=30, title="", text_scaling = 0.75,
                                   use_asv = FALSE,
                                   taxo_level= genus,
                                   taxo_level_fill = supergroup,
                                   taxo_colors_fill = global$supergroup_color)

gg_stations <- stations$gg + 
  ylab("% of samples where detected") +
  ggtitle("Most frequent genera")
  

asvs <- asv_set_V4$df %>% 
  select(asv_code, supergroup, genus) %>% 
  distinct() %>% 
  group_by(supergroup, genus) %>% 
  count() %>% 
  mutate(n_reads = n) %>%
  filter(!str_detect(genus, genus_removed)) %>% 
  phyloseq_long_bargraph (n_bars=30, title="", text_scaling = 0.75,
                                   use_asv = FALSE,
                                   taxo_level= genus,
                                   taxo_level_fill = supergroup,
                                   taxo_colors_fill = global$supergroup_color) 
gg_asvs <- asvs$gg + 
  ylab("Number of ASVs") +
  ggtitle("Genera with most cASVs") 



# --------------------------------------
layout <- "
AACC
AACC
BBDD
BBDD
"


fig_bars_genus <- gg_reads + gg_stations + gg_asvs + guide_area()  +
  plot_layout(guides = "collect", design = layout) +
  plot_annotation(tag_levels = 'A') & 
  theme(plot.tag = element_text(size = 30)) 

fig_bars_genus
```


### Correlations

```{r}

genus <- reads$df %>% 
  left_join(rename(stations$df, n_stations = n_reads)) %>% 
    left_join(rename(asvs$df, n_asvs = n_reads)) 

ggplot(data = genus) + 
  geom_point(aes(x=n_asvs, y=n_reads, color = supergroup), size = 3) + 
  scale_x_log10() + scale_y_log10() +
  scale_color_manual(values = global$supergroup_colors)

ggplot(data = genus) + 
  geom_point(aes(x=n_asvs, y=n_stations, color = supergroup), size = 3) + 
  scale_x_log10() + scale_y_log10() +
  scale_color_manual(values = global$supergroup_colors)
```

## Analysis at species level

### Species found in different environements - Upset

```{r fig.height=12, fig.width=12}
df <- asv_set_V4$df %>% 
  select(asv_code, ecosystem) %>% 
  distinct() %>% 
  mutate(n=1) %>% 
  pivot_wider(names_from = ecosystem, values_fill = 0, values_from = n) %>% 
  data.frame()


fig_upset_asv <- UpSetR::upset(df, 
                                nsets = 5,
                                empty.intersections = "on", 
                                nintersects = 14,
                                number.angles = 30, 
                                point.size = 10, 
                                line.size = 4, 
                                mainbar.y.label = "ASV number", 
                                sets.x.label = "ASVs", 
                                order.by = "freq", 
                                text.scale = c(3), 
                                mb.ratio = c(0.6, 0.4))



fig_upset_asv

```


## Analysis at ASV level

### Most abundant ASV

```{r}

reads <- asv_set_V4$df %>% 
  mutate(n_reads = n_reads_pct/n_reads_pct_V4*100) %>%
  # filter(!str_detect(genus, "_X")) %>% 
  phyloseq_long_bargraph (n_bars=30, title="", text_scaling = 0.75,
                                   use_asv = TRUE,
                                   taxo_level= species,
                                   taxo_level_fill = supergroup,
                                   taxo_colors_fill = global$supergroup_color)
 gg_reads <- reads$gg  + 
  ylab("% of total number of reads") +
  ggtitle("Most abundant cASVs") +
  theme_classic() +
  theme(legend.position="none")
 
 
  

# --------------------------------------
  

stations <- asv_set_V4$df %>% 
  select(file_code, kingdom, supergroup, family, species, asv_code) %>% 
  distinct() %>% 
  group_by(kingdom, supergroup, family, species, asv_code) %>% 
  count() %>% 
  mutate(n_reads = n/n_samples_V4*100) %>%
  # filter(!str_detect(genus, "_X")) %>% 
  phyloseq_long_bargraph (n_bars=30, title="", text_scaling = 0.75,
                                   use_asv = TRUE,
                                   taxo_level= genus,
                                   taxo_level_fill = supergroup,
                                   taxo_colors_fill = global$supergroup_color)
gg_stations <- stations$gg + 
  ylab("% of samples where detected") +
  ggtitle("Most frequent cASVs")  +
  theme_classic() +
  theme(legend.position="bottom") 
  
  

# --------------------------------------

asvs <- reads$df %>% 
  left_join(rename(stations$df, n_stations = n_reads)) 

fig_correl_asv <- ggplot(data = asvs) + 
  geom_point(aes(x=n_stations, y=n_reads, color = supergroup), size = 1.5) + 
  # scale_x_log10() +
  scale_y_log10(limits = c(0.01, 2)) +
  # ylim(0.01, 2) +
  scale_color_manual(values = global$supergroup_colors) +
  xlab("% of samples where detected") +
  ylab("% of total number of reads") +
  # geom_smooth(aes(x=n_stations, y=n_reads), method = "lm") +
  guides(color = "none") +
  theme_bw() +
  ggrepel::geom_text_repel(data = filter(asvs, n_stations > 2), 
                           aes(x=n_stations, y=n_reads, label = str_sub(bar_label, 12)),
                           size = 2)

fig_correl_asv

# --------------------------------------

layout <- "
AAAA
AAAA
BBBB
BBBB
"


fig_bars_asv <- gg_reads + gg_stations  + guide_area() +
  plot_layout(design = layout) +
  plot_annotation(tag_levels = 'A') & 
  theme(plot.tag = element_text(size = 30))

fig_bars_asv
```

### ASV found in different environements - Upset

```{r fig.height=12, fig.width=12}
df <- asv_set_V4$df %>% 
  select(asv_code, ecosystem) %>% 
  distinct() %>% 
  mutate(n=1) %>% 
  pivot_wider(names_from = ecosystem, values_fill = 0, values_from = n) %>% 
  data.frame()




fig_upset_asv <- UpSetR::upset(df, 
                                nsets = 5,
                                empty.intersections = "on", 
                                nintersects = 8,
                                number.angles = 30, 
                                point.size = 12, 
                                line.size = 4, 
                                mainbar.y.label = "cASV number", 
                                sets.x.label = "cASVs", 
                                order.by = "freq", 
                                text.scale = c(5), 
                                mb.ratio = c(0.7, 0.3))



fig_upset_asv

```

## Barplots at supergroup level

### By environment

```{r}

supergroup_to_remove <- c("Apusozoa", "Eukaryota_X", "Excavata", "Other")

df <- asv_set_V4$df %>% 
  select(file_code, ecosystem, supergroup, ecological_function, trophic_group, n_reads_pct) %>% 
  mutate(supergroup = fct_other(supergroup,
                                drop = supergroup_to_remove,
                                other_level = "Other"),
         ecosystem = fct_rev(ecosystem))

supergroup_colors <- global$supergroup_colors[setdiff(names(global$supergroup_colors), supergroup_to_remove)]

supergroup_colors <- c(supergroup_colors, "Other" = "#000000")

ecological_function_colors <- global$ecological_function_colors[setdiff(names(global$ecological_function_colors), "metazoans")]

trophic_group_colors <- global$trophic_group_colors[setdiff(names(global$trophic_group_colors), c("metazoan", "fungi"))]


fig_barplot_env_taxo <- barplot(df, "ecosystem","taxonomy" , "supergroup" )

plotly::ggplotly(fig_barplot_env_taxo)

fig_barplot_env_ecological_function <- barplot(df, "ecosystem","ecological_function" , "supergroup" )

plotly::ggplotly(fig_barplot_env_ecological_function)

fig_barplot_env_trophic_group <- barplot(df, "ecosystem","trophic_group" , "supergroup" )

plotly::ggplotly(fig_barplot_env_trophic_group)

```

### By latitude for marine samples only

```{r}

supergroup_to_remove <- c("Apusozoa", "Eukaryota_X", "Excavata", "Other")

df <- asv_set_V4$df %>% 
  filter(ecosystem %in% c("oceanic", "coastal")) %>% 
  select(file_code, latitude, supergroup, ecological_function, n_reads_pct) %>% 
  mutate(supergroup = fct_other(supergroup,
                                drop = supergroup_to_remove,
                                other_level = "Other")) 

supergroup_colors <- global$supergroup_colors[setdiff(names(global$supergroup_colors), supergroup_to_remove)]

supergroup_colors <- c(supergroup_colors, "Other" = "#000000")

ecological_function_colors <- global$ecological_function_colors[setdiff(names(global$ecological_function_colors), "metazoans")]


fig_barplot_lat_taxo <- barplot(df, "latitude","taxonomy" , "supergroup" ) +
  ylab("Latitude") 

plotly::ggplotly(fig_barplot_lat_taxo)

fig_barplot_lat_traits <- barplot(df, "latitude","ecological_function" , "supergroup" )  +
  ylab("Latitude")

plotly::ggplotly(fig_barplot_lat_traits) 

fig_barplot_lat <-fig_barplot_lat_taxo / fig_barplot_lat_traits +
  plot_annotation(tag_levels = 'A')

```

## Diversity analyis

### By environment

#### Make phyloseq

All

phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 33325 taxa and 2915 samples ]
sample_data() Sample Data:       [ 2915 samples by 9 sample variables ]
tax_table()   Taxonomy Table:    [ 33325 taxa by 8 taxonomic ranks ]

Ouliers removed (24 samples)
phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 33219 taxa and 2891 samples ]
sample_data() Sample Data:       [ 2891 samples by 9 sample variables ]
tax_table()   Taxonomy Table:    [ 33219 taxa by 8 taxonomic ranks ]


```{r}
ps <- make_phyloseq(asv_set_V4$samples, asv_set_V4$df, asv_set_V4$fasta)

ps

qs::qsave(ps, "../data/ps_V4_env.qs")

# Remove outliers

outliers <- qs::qread("../data/ps_V4_outliers.qs")

samples <- asv_set_V4$samples %>% 
  filter(!(file_code %in% outliers$file_code))

df <- asv_set_V4$df %>% 
  filter(!(file_code %in% outliers$file_code))

fasta <-asv_set_V4$fasta %>% 
  filter(asv_code %in% df$asv_code)

ps <- make_phyloseq(samples, df, fasta)

ps

qs::qsave(ps, "../data/ps_V4_env_outliers_removed.qs")


```


#### Alpha
```{r}


  
  

fig_alpha_env <- ps_alpha(ps, measures = c("Shannon"), 
                          x="ecosystem" , color="ecosystem", shape="fraction_name") +
                 xlab("Ecosystem") + ylab("Diversity")

fig_alpha_env_significance <-microbiomeSeq::plot_anova_diversity(ps, method = c("shannon"),
                                    grouping_column =  "ecosystem",pValueCutoff=0.05)

fig_alpha_env

fig_alpha_env_significance

```

#### Beta

* Need to remove outliers

Call:
metaMDS(comm = veganifyOTU(physeq), distance = distance, autotransform = TRUE,      parallel = 16) 

global Multidimensional Scaling using monoMDS

Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
Distance: bray 

Dimensions: 2 
Stress:     0.06828168 
Stress type 1, weak ties
No convergent solutions - best solution after 32 tries
Scaling: centring, PC rotation, halfchange scaling 
Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’

```{r}

# ps_ordinate <- phyloseq::ordinate(ps, "NMDS", "bray")

ps_ordinate <- qs::qread( "../data/ps_ordinate_V4_env_NMDS_bray_outliers_removed.qs")

ps_ordinate

# ps_ordinate <- qs::qread( "../data/ps_ordinate_V4_env_PCA.qs")

# Find outliers from NMDS

# points <- data.frame(ps_ordinate[["points"]]) 
# 
# outliers <- points %>% 
#   filter((MDS1 < -0.01) | (MDS1 > 0.01) |( MDS2 < -0.01) | (MDS2 > 0.01)) %>% 
#   tibble::rownames_to_column(var = "file_code")
# 
# qs::qsave(outliers, "../data/ps_V4_outliers.qs")



fig_beta_env <- ps_beta(ps, ps_ordinate, 
                    color_samples="ecosystem", 
                    shape_samples = NULL,
                    color_taxa = "supergroup")

fig_beta_env_samples <- fig_beta_env$gg_samples
fig_beta_env_taxa <- fig_beta_env$gg_taxa

plotly::ggplotly(fig_beta_env_samples)
plotly::ggplotly(fig_beta_env_taxa)
```



### By latitude for only the marine stuff

```{r}

df <- asv_set_V4$df %>% 
  filter(ecosystem %in% c("oceanic", "coastal")) 
fasta <- asv_set_V4$fasta %>% 
  filter(asv_code %in% df$asv_code)
samples <- asv_set_V4$samples %>% 
  filter(file_code %in% df$file_code)

ps <- make_phyloseq(samples, df, fasta)

ps

qs::qsave(ps, "../data/ps_V4_marine.qs")

fig_alpha_lat <- ps_alpha(ps, measures = c("Shannon"), 
                          x="latitude" , color="ecosystem", shape="fraction_name",
                          discretize = TRUE) +
                 xlab("Latitude") + ylab("Diversity")

```

# BLAST (usearch) analysis

## Read usearch table

```{r}
metapr2_db <- dvutils::db_info("metapr2_google")
metapr2_db_con <- dvutils::db_connect(metapr2_db)

usearch <- tbl(metapr2_db_con, "metapr2_asv_vsearch_pr2") %>% 
  collect()

 dvutils::db_disconnect(metapr2_db_con)
 
 usearch <- usearch %>% 
   mutate(asv_code = str_sub(hash_value, 1, 10)) %>% 
   select(-hash_value) %>% 
   tidyr::separate(pr2_label, into = c("pr2_accession", NA, NA, "seq_name",str_c(global$taxo_levels[1:8],"_usearch" )), sep = "[|]") %>% 
   mutate(pr2_sequence_type = case_when(str_detect(seq_name, "strain") ~ "cultured",
                              TRUE ~ "environmental"))
 
 usearch_summary <- usearch %>% 
   group_by(asv_code) %>% 
   slice_max(order_by = pr2_sequence_type, n=1, with_ties = FALSE) %>% 
   ungroup() 
 
```

## Merge with ASV

```{r}
df_usearch <- asv_set_V4$df %>% 
  select(any_of(global$taxo_levels), ecological_function, trophic_group, file_code, n_reads, n_reads_pct, latitude, ecosystem) %>% 
  left_join(select(usearch_summary, asv_code, pr2_sequence_type, pct_id, contains("_usearch")), by = "asv_code") 
```

## Distribution based on similarity by supergroup

```{r fig.height=10, fig.width=10}

supergroup_to_remove <- c("Apusozoa", "Eukaryota_X", "Excavata", "Other") # less than 100 ASVs
# supergroup_to_remove <- c("Eukaryota_X", "Excavata", "Other")

supergroup_colors <- global$supergroup_colors[setdiff(names(global$supergroup_colors), supergroup_to_remove)]

# supergroup_colors <- c(supergroup_colors, "Other" = "#000000")

df <- df_usearch %>% 
  select(asv_code, supergroup, pct_id, pr2_sequence_type) %>% 
  distinct() %>% 
  filter(pct_id > 0, 
         !(supergroup %in% supergroup_to_remove)) 

  # mutate(supergroup = fct_other(supergroup,
  #                               drop = supergroup_to_remove,
  #                               other_level = "Other")
  # ) 

table_pct_id <- df %>% 
  group_by(supergroup) %>% 
  summarize(pct_id_mean = mean(pct_id), n = n()) %>% 
  arrange(desc(pct_id_mean))

table_pct_id

df <- left_join(df, table_pct_id)

fig_pct_id_density <- ggplot(df, aes(x=pct_id, fill = supergroup)) + 
  geom_density(alpha=1) +
  # facet_wrap( ~ pr2_sequence_type) 
  facet_wrap( ~ supergroup, ncol = 3, scales = "free_y") +
  scale_fill_manual(values = supergroup_colors)+
  # ylim(0, 0.5) + 
  xlim(80,100) +
  theme_bw()  
fig_pct_id_density 

fig_pct_id_ridges <- ggplot(df, aes(x=pct_id, y = reorder(supergroup, pct_id_mean), fill = supergroup)) + 
  ggridges::geom_density_ridges(alpha=0.8, scale = 4, quantile_lines = TRUE, quantiles = 2, bandwidth = 1) +
  # facet_wrap( ~ pr2_sequence_type) 
  # facet_wrap( ~ supergroup, ncol = 3, scales = "free_y") +
  scale_fill_manual(values = supergroup_colors)+
  # ylim(0, 0.5) + 
  xlim(80,100) +
  ggridges::theme_ridges()  +
  ylab("") +
  xlab("% identity with PR2 sequences")


# plotly::ggplotly(fig_pct_id_density)

fig_pct_id_ridges  


fig_pct_id_violin <- ggplot(df, aes(y=pct_id, x = supergroup)) + 
  geom_violin() +
  # ggforce::geom_sina(aes(color=.data[[color]], shape =.data[[shape]] )) +
  ggforce::geom_sina() +
  geom_boxplot(color = "grey", alpha = 0, width=0.5) +
  # scale_color_manual(values = global$supergroup_colors)+
  # ylim(0, 0.5) + xlim(70,100) +
  theme_bw()  +
  coord_flip()

fig_pct_id_violin



```

## Distribution based on similarity by ecolgical function

```{r fig.height=10, fig.width=10}

df <- df_usearch %>% 
  select(asv_code, ecological_function, pct_id, pr2_sequence_type) %>% 
  distinct() %>% 
  filter(!is.na(ecological_function))

  # mutate(supergroup = fct_other(supergroup,
  #                               drop = supergroup_to_remove,
  #                               other_level = "Other")
  # ) 

table_pct_id <- df %>% 
  group_by(ecological_function) %>% 
  summarize(pct_id_mean = mean(pct_id), n = n()) %>% 
  arrange(desc(pct_id_mean)) 

table_pct_id

df <- left_join(df, table_pct_id)

fig_pct_id_ridges_ecological_function <- ggplot(df, aes(x=pct_id, y = reorder(ecological_function, pct_id_mean), fill = ecological_function)) + 
  ggridges::geom_density_ridges(alpha=0.8, scale = 4, quantile_lines = TRUE, quantiles = 2, bandwidth = 1) +
  # facet_wrap( ~ pr2_sequence_type) 
  # facet_wrap( ~ supergroup, ncol = 3, scales = "free_y") +
  scale_fill_manual(values = ecological_function_colors)+
  # ylim(0, 0.5) + 
  xlim(80,100) +
  ggridges::theme_ridges()  +
  ylab("") +
  xlab("% identity with PR2 sequences")


# plotly::ggplotly(fig_pct_id_density)

fig_pct_id_ridges_ecological_function  

```

## Distribution based on similarity by environements

```{r fig.height=10, fig.width=10}

supergroup_to_remove <- c("Apusozoa", "Eukaryota_X", "Excavata", "Other")
# supergroup_to_remove <- c("Eukaryota_X", "Excavata", "Other")

supergroup_colors <- global$supergroup_colors[setdiff(names(global$supergroup_colors), supergroup_to_remove)]

# supergroup_colors <- c(supergroup_colors, "Other" = "#000000")

df <- df_usearch %>% 
  select(asv_code, ecosystem, pct_id, supergroup) %>% 
  distinct() %>% 
  filter(pct_id > 0, 
         !(supergroup %in% supergroup_to_remove))  


  # mutate(supergroup = fct_other(supergroup,
  #                               drop = supergroup_to_remove,
  #                               other_level = "Other")
  # )

fig_pct_id_env_density <- ggplot(df, aes(x=pct_id)) + 
  geom_density(alpha=1, fill="grey") +
  # facet_wrap( ~ pr2_sequence_type) 
  facet_wrap( ~ ecosystem, ncol = 2, scales = "free_y") +
  # ylim(0, 0.5) + 
  xlim(80,100) +
  theme_bw()  

fig_pct_id_env_density

fig_pct_id_env_ridges <- ggplot(df, aes(x=pct_id, y = ecosystem, fill = ecosystem)) + 
  ggridges::geom_density_ridges(alpha=0.8, scale = 2, quantile_lines = TRUE, quantiles = 2, bandwidth = 1) +
  # facet_wrap( ~ pr2_sequence_type) 
  # facet_wrap( ~ supergroup, ncol = 3, scales = "free_y") +
  # scale_fill_manual(values = supergroup_colors)+
  # ylim(0, 0.5) + 
  xlim(80,100) +
  ggridges::theme_ridges()  +
  scale_fill_viridis_d() +
  ylab("") +
  xlab("% identity with PR2 sequences") + 
  theme(legend.position = "none")


# plotly::ggplotly(fig_pct_id_density)

fig_pct_id_env_ridges  

fig_pct_id_env_jitter <- ggplot(df, aes(y=pct_id, x = ecosystem)) + 
  # geom_violin() +
  # ggforce::geom_sina(aes(color=.data[[color]], shape =.data[[shape]] )) +
  # ggforce::geom_sina() +
  geom_boxplot(color = "black", alpha = 0, width=0.5) +
  geom_jitter(aes(color = supergroup), width = 0.1) +
  scale_color_manual(values = supergroup_colors )+
  # ylim(0, 0.5) + xlim(70,100) +
  theme_bw()  +
  coord_flip()

fig_pct_id_env_jitter

# plotly::ggplotly(fig_pct_id_env_jitter)


```



# Combine figures together

## Barplot

```{r}

fig_barplot_env <-(fig_barplot_env_taxo + 
  theme(legend.position="top", legend.box = "horizontal") +
  guides(fill = guide_legend(title.position = "top",
                             title = "Supergroup",
                              title.hjust = 0.5)) +
  
  fig_barplot_env_ecological_function +
  theme(legend.position="top", legend.box = "horizontal")+
  guides(fill = guide_legend(title.position = "top",
                             title = "Ecological function", 
                             title.hjust = 0.5)) ) /
  (fig_alpha_env +
       xlab("") + ylab("Shannon index") +
      theme(
      strip.background = element_blank(),
      strip.text.x = element_blank()
      )
    + fig_pct_id_env_ridges  + 
  theme(legend.position = "none") ) +
  plot_annotation(tag_levels = 'A') &
  theme(plot.tag = element_text(size = 30))   

fig_barplot_env

# ---------------------------------------------

fig_barplot_lat <-fig_barplot_lat_taxo / 
  fig_barplot_lat_traits /
  (fig_alpha_lat +
       xlab("Latitude") + ylab("Shannon index") +
      theme(
      strip.background = element_blank(),
      strip.text.x = element_blank()
      )
   ) +
  plot_annotation(tag_levels = 'A')


# ---------------------------------------------


fig_alpha = (fig_alpha_env +
                 xlab("") + ylab("Diversity") +
                theme(
                strip.background = element_blank(),
                strip.text.x = element_blank()
                )
             )  / 
            (fig_alpha_lat  +
                 xlab("Latitude") + ylab("Diversity") +
                theme(
                strip.background = element_blank(),
                strip.text.x = element_blank()
                )
             ) +
            plot_annotation(tag_levels = 'A')

fig_alpha

```


## % Similarity

```{r}

fig_pct_id <- fig_pct_id_ridges + fig_pct_id_ridges_ecological_function +
  plot_annotation(tag_levels = 'A') & 
  theme(plot.tag = element_text(size = 12)) + 
  theme(legend.position = "none") 
  # guides(linetype = "none", fill = guide_legend(override.aes = list(linetype=0), nrow = 3)) &  
  # theme(legend.position="top", legend.box = "horizontal")
fig_pct_id

```

# Export figures

```{r, eval=TRUE}

ggsave(plot= fig_treemap , filename=path_fig("fig_treemap.pdf"), 
       width = 20 , height = 12, scale=1, units="cm", useDingbats=FALSE) 

ggsave(plot= fig_treemap_all , filename=path_fig("fig_treemap_all.pdf"), 
       width = 20 , height = 12, scale=1, units="cm", useDingbats=FALSE) 

ggsave(plot= fig_bars_genus , filename=path_fig("fig_bars_genus.pdf"), 
       width = 16 , height = 12, scale=2.5, units="cm", useDingbats=FALSE)

ggsave(plot= fig_bars_asv , filename=path_fig("fig_bars_asvs.pdf"), 
       width = 15, height = 25, scale=1.2, units="cm", useDingbats=FALSE)

ggsave(plot= fig_correl_asv , filename=path_fig("fig_correl_asvs.pdf"), 
       width = 12, height = 12, scale=1.0, units="cm", useDingbats=FALSE)

ggsave(plot= fig_barplot_env , filename=path_fig("fig_barplot_env.pdf"), 
       width = 22, height = 15, scale=1.5, units="cm", useDingbats=FALSE)

ggsave(plot= fig_barplot_lat , filename=path_fig("fig_barplot_lat.pdf"), 
       width = 15, height = 22, scale=1.5, units="cm", useDingbats=FALSE)

ggsave(plot= fig_alpha_env_significance , filename=path_fig("fig_alpha_env_significance.pdf"), 
       width = 12, height = 18, scale=1.5, units="cm", useDingbats=FALSE)

ggsave(plot= fig_beta_env_samples , filename=path_fig("fig_beta_env.pdf"), 
       width = 18, height = 12, scale=1.5, units="cm", useDingbats=FALSE)

ggsave(plot= fig_pct_id , filename=path_fig("fig_pct_id.pdf"), 
       width =20, height = 12, scale=2, units="cm", useDingbats=FALSE)

pdf(path_fig("fig_upset_asv.pdf"), width = 30, height = 12, pointsize = 10,  useDingbats = FALSE)
fig_upset_asv
dev.off()

# ggsave(plot= fig_year , filename=path_fig("fig_year.tiff"), 
#        width = 16 , height = 8, scale=1.8, units="cm", compression = "lzw") 


```

# Clean some stuff

```{r}
rm(ps)

rm(asv_set)


```

